integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/InVMKafkaCluster.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.zookeeper.server.ServerCnxnFactory;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.testcluster.ClusterConfig.ConfigHolder;
+import io.kroxylicious.proxy.testcluster.ClusterConfig.KafkaEndpoints;
+import kafka.server.KafkaConfig;
+import kafka.server.KafkaRaftServer;
+import kafka.server.KafkaServer;
+import kafka.server.Server;
+import kafka.tools.StorageTool;
+import scala.Option;
+
+public class InVMKafkaCluster implements Cluster {
+    private static final Logger LOGGER = LoggerFactory.getLogger(InVMKafkaCluster.class);
+
+    private final ClusterConfig clusterConfig;
+    private final Path tempDirectory;
+    private final ServerCnxnFactory zooFactory;
+    private final ZooKeeperServer zooServer;
+    private List<Server> servers;
+    private List<String> bootstraps = new ArrayList<>();
+
+    public InVMKafkaCluster(ClusterConfig clusterConfig) {
+        this.clusterConfig = clusterConfig;
+        try {
+            tempDirectory = Files.createTempDirectory("kafka");
+
+            var numPorts = clusterConfig.getBrokersNum() * (clusterConfig.isKraftMode() ? 3 : 2) + (clusterConfig.isKraftMode() ? 0 : 1);
+            LinkedList<Integer> ports = Utils.preAllocateListeningPorts(numPorts).collect(Collectors.toCollection(LinkedList::new));
+
+            final Supplier<KafkaEndpoints.Endpoint> zookeeperEndpointSupplier;
+            if (!clusterConfig.isKraftMode()) {
+                var zookeeperPort = ports.pop();
+
+                zooFactory = ServerCnxnFactory.createFactory(new InetSocketAddress("localhost", zookeeperPort), 1024);
+
+                var zoo = tempDirectory.resolve("zoo");
+                var snapshotDir = zoo.resolve("snapshot");
+                var logDir = zoo.resolve("log");
+                snapshotDir.toFile().mkdirs();
+                logDir.toFile().mkdirs();
+
+                zooServer = new ZooKeeperServer(snapshotDir.toFile(), logDir.toFile(), 500);
+                zookeeperEndpointSupplier = () -> new KafkaEndpoints.Endpoint("localhost", zookeeperPort);
+            }
+            else {
+                zooFactory = null;
+                zooServer = null;
+                zookeeperEndpointSupplier = null;
+            }
+
+            Supplier<KafkaEndpoints> kafkaEndpointsSupplier = () -> new KafkaEndpoints() {
+                final List<Integer> clientPorts = ports.subList(0, clusterConfig.getBrokersNum());
+                final List<Integer> interBrokerPorts = ports.subList(clusterConfig.getBrokersNum(), 2 * clusterConfig.getBrokersNum());
+                final List<Integer> controllerPorts = ports.subList(clusterConfig.getBrokersNum() * 2, ports.size());
+
+                @Override
+                public EndpointPair getClientEndpoint(int brokerNum) {
+                    var port = clientPorts.get(brokerNum);
```

Comment:
Could we rename the variable `brokerNum` to something like `brokerId`? It is really confusing when putting together with `brokersNum`

Reply:
Done.